### PR TITLE
Add README guide

### PR DESCRIPTION
**Pull request by ddbeck**
This adds some guidelines for writing and setting metadata for READMEs that will play nicely with PyPI. This should cover a lot of the things discussed in #210.

### CODE REVIEW

File: source/guides/making-a-pypi-friendly-readme.rst [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+For your README to display properly on PyPI, choose a markup language supported by PyPI.
+Formats supported by `PyPI's README renderer <https://github.com/pypa/readme_renderer>`_ are:
+
+* plain text
+* `reStructuredText <http://docutils.sourceforge.net/rst.html>`_
+* Markdown (`GitHub Flavored Markdown <https://github.github.com/gfm/>`_ by default,
+  or `CommonMark <http://commonmark.org/>`_)
+
+It's customary to save your README file in the root of your project, in the same directory as your ``setup.py`` file.
```
Comment (pradyunsg) [inline source/guides/making-a-pypi-friendly-readme.rst]: I think saying "setup.py or pyproject.toml" here would be nice.

Comment (ddbeck) [reply source/guides/making-a-pypi-friendly-readme.rst]: This is a good idea, @pradyunsg ! I'll rework this a little to mention `pyproject.toml`.

Even better though would be to include a first-class `pyproject.toml` example. Do you know of any projects I could look at that specify a README and markup in their `pyproject.toml`?

Comment (ncoghlan) [reply source/guides/making-a-pypi-friendly-readme.rst]: I think we want to hold off on pushing `pyproject.toml` too heavily until there's a version of `pip` available with initial PEP 517 support for pluggable build backends.

Then we can add a sample `flit` project with only `pyproject.toml` (assuming flit supports getting its settings from there).

Comment (ddbeck) [reply source/guides/making-a-pypi-friendly-readme.rst]: OK, well, I still reworked things to leave room for other ways of specifying—it shouldn't require too much rewriting to add `pyproject.toml` when it becomes a practical option.

File: source/guides/making-a-pypi-friendly-readme.rst [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+* create a README in a PyPI-friendly format
+* include your README in your package so it appears on PyPI
+* find resources for writing README content that will help your users
+
+
+Creating a README file
+----------------------
+
+README files are often named ``README`` or ``README.txt`` for plain-text READMEs,
+or :samp:`README.{extension}` where *extension* indicates a markup language
```
Comment (jonparrott) [inline source/guides/making-a-pypi-friendly-readme.rst]: why not just enumerate all of them? We only support 3 different content types `README.txt`, `README.md`, and `README.rst`.

File: source/guides/making-a-pypi-friendly-readme.rst [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+Write your README file
+----------------------
+
+Your README file can contain whatever you want, but it's common to include information that helps your reader understand:
+
+* what your project is and how it can help them
+* how to install or use your project
+* where to go to contribute or get help
+
```
Comment (jonparrott) [inline source/guides/making-a-pypi-friendly-readme.rst]: I would love a short example here.

File: source/guides/making-a-pypi-friendly-readme.rst [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+Write your README file
+----------------------
+
+Your README file can contain whatever you want, but it's common to include information that helps your reader understand:
+
+* what your project is and how it can help them
+* how to install or use your project
+* where to go to contribute or get help
```
Comment (jonparrott) [inline source/guides/making-a-pypi-friendly-readme.rst]: The readme should also contain licensing information.

File: source/guides/making-a-pypi-friendly-readme.rst [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+.. seealso::
+
+   * :any:`metadata_description`
+   * :any:`metadata_description_content_type`
+
+For example, to set these values in a package's :file:`setup.py` file,
+use ``setup()``'s ``long_description`` and ``long_description_content_type``.
+
+Set the value of ``long_description`` to the contents of the README file itself.
```
Comment (jonparrott) [inline source/guides/making-a-pypi-friendly-readme.rst]: suggestion: make it clear **not** to use the readme file name:

```
Set the value of ``long_description`` to the contents of the README file itself, not to the file name of the README file.
```

File: source/guides/making-a-pypi-friendly-readme.rst [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+   from os import path
+   this_directory = path.abspath(path.dirname(__file__))
+   with open(path.join(this_directory, 'README.md'), encoding='utf-8') as f:
+       long_description = f.read()
+
+   setup(
+       name='an_example_package',
+       # other arguments omitted
+       long_description=long_description,
+       long_description_content_type='text/markdown; variant=gfm'
```
Comment (jonparrott) [inline source/guides/making-a-pypi-friendly-readme.rst]: drop the variant here, as GFM is the default.

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 4 above)
Comment (ddbeck) [reply source/guides/making-a-pypi-friendly-readme.rst]: I don't want to go overboard about details of README content and I'm not sure there's a brief way to describe describing a license in a README file.

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 2 above)
Comment (ddbeck) [reply source/guides/making-a-pypi-friendly-readme.rst]: Because other formats exist even if we don't support them yet?

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 3 above)
Comment (ddbeck) [reply source/guides/making-a-pypi-friendly-readme.rst]: I worry that even a short example would be long enough to dominate the page. I'd rather link to resources and examples elsewhere.

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 6 above)
Comment (ddbeck) [reply source/guides/making-a-pypi-friendly-readme.rst]: OK, I'll change this.

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 5 above)
Comment (ddbeck) [reply source/guides/making-a-pypi-friendly-readme.rst]: OK, I can fix this up.

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 4 above)
Comment (di) [reply source/guides/making-a-pypi-friendly-readme.rst]: I agree, this is what the license classifiers and the `License` metadata field are for.

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 2 above)
Comment (jonparrott) [reply source/guides/making-a-pypi-friendly-readme.rst]: This is written as if in the abstract. This guide should focus on the *actual*:

```
READMEs for Python projects are generally named ``README``, ``README.rst``, or ``README.md``.
```

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 3 above)
Comment (jonparrott) [reply source/guides/making-a-pypi-friendly-readme.rst]: That's fine considering you link to a sample.

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 4 above)
Comment (jonparrott) [reply source/guides/making-a-pypi-friendly-readme.rst]: There absolutely is a brief way of describing it:

```
* the name of the license your project is under.
```

See [here](https://github.com/GoogleCloudPlatform/google-auth-library-python#license) for example.

It's so incredibly critical that this information is in the project description as many packages can omit the license metadata.

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 2 above)
Comment (ddbeck) [reply source/guides/making-a-pypi-friendly-readme.rst]: OK.

File: source/guides/making-a-pypi-friendly-readme.rst (hunk 4 above)
Comment (ncoghlan) [reply source/guides/making-a-pypi-friendly-readme.rst]: Aside from "It was the early 2000s and licensing discipline wasn't anywhere near as well understood back then as some of understand it now", PyPI also doesn't require explicit licence metadata because it's the Terms of Service that set the minimum licensing terms (i.e. the ability to make verbatim copies of the uploaded files, even if you're not supposed to do anything else with them).

That said, I agree with @ddbeck that it makes sense to remain silent on the topic of including licensing info in README files for now, especially since platforms like GitHub will extract and display the license from your repo anyway. (Clarity in open source licensing for uploaded projects is also an area that could use a whole lot of improvements in general, and a single bullet point here isn't going to resolve those issues)

Comment (jonparrott) [reply source/guides/making-a-pypi-friendly-readme.rst]: I don't mean to be a jerk, and I'm happy to be outvoted here, but please here my last plea to include this here. I consider this critically important to OSS, so, please, hear me out just once more.

---

LIcensing is like security. It should be done *in depth*. There's a reason Apache and friends tell you to put a *license header* on every file. It's to reduce the amount of ambiguity around what license things are under. Naming the license in the README is just another avenue to reduce ambiguity (I *never* suggested it should be the only thing you do). Naming the license *is sufficient* for this task (it's akin to the license header in a source file). Keep in mind that in lots of situations it's entirely possible for the package metadata to be lost in one way or another (repackaging, etc.).

---

I also don't understand why there is so much push back on this item. I would argue that there are lots of projects that *will not* accept contributions, lots of projects that aren't meant to be installed broadly, and lots of projects that are made for just one person. This sort of argument can be used to just remove all of these bullet points. But critically, every package uploaded to the package index should indicate, somehow, even if it's just by name, what the terms of use are.

Comment (ddbeck) [reply source/guides/making-a-pypi-friendly-readme.rst]: Thanks for your comment @jonparrott and your patience while I got back to you on this.

To go meta for a moment, you asked why there was push back here. I can't speak for anyone else, but I put a lot of care into this doc and I had good reasons for pretty much everything I wrote from the outset. And I have bit of a personal attachment to the topic generally, since "Daniel cares a lot about READMEs" seems to be my personal brand of late. What can I say? I care about this.

Now for the actual issue: I see where you're coming from. I agree with you that licensing is important. But I don't think that providing licensing information to users, in isolation, is a primary goal of README files. Nor do I think licensing represents a major problem for README authors: when I surveyed READMEs for a conference talk, I found that remembering to name your project and link to the repo or homepage is a more significant and pervasive failure on the part of README authors than licensing. 😬

But I'm also uncomfortable with implying that naming a license is significant and meaningful (by itself, outside a bigger discussion on software licensing). I believe that approach contributes to a ritualistic view of copyright (and law broadly) in which people believe [magic words will bind others' actions](https://www.pixsy.com/does-no-copyright-infringement-intended-have-any-legal-significance/). It feels like a sort of anti-civics lesson.

I have to say though, I really like this general idea of licensing in-depth. I would _love_ to see a doc on how to properly license, top to bottom, a Python package (and maybe even address some Python specific examples—like what's the relationship between `import somelib` and its license? I wish I knew!). I think my ideal outcome would be to link to that page from this one, rather than trying to cover a tiny slice of it here.

Comment (jonparrott) [reply source/guides/making-a-pypi-friendly-readme.rst]: > I would love to see a doc on how to properly license, top to bottom, a Python package (and maybe even address some Python specific examples—like what's the relationship between import somelib and its license? I wish I knew!). I think my ideal outcome would be to link to that page from this one, rather than trying to cover a tiny slice of it here.

I would love that, want to file a separate issue?

---

Back to the topic at hand - Let me rephrase your argument *against* adding this as an argument against adding one of the items you already have in this list:

>  But I don't think that providing *contribution* information to users, in isolation, is a primary goal of README files. Nor do I think *accepting contributions* represents a major problem for README authors.

My argument is this: either this section should contain *just the bare minimum* which you identified through survey as just "the project name and homepage" or it we should amend this list to include the name of the license. If it's going to be an arbitrary list of common items, license should be there. If it's going to be the bare minimum, then none of these other things should be there.

Comment (jonparrott) [reply source/guides/making-a-pypi-friendly-readme.rst]: Also want to offer a reasonable bridge here: We could go the bare minimum route here and dp something to this effect:

> Your README should at minimum contain your project's name and a link to your project's homepage where users can find out more information. For a information on writing a more comprehensive project README, we recommend using [readme-checklist](https://github.com/ddbeck/readme-checklist/blob/master/checklist.md).

Your README checklist project more than satisfies my desires in terms of documenting the license and I am more than happy to promote it here.

Comment (jonparrott) [reply source/guides/making-a-pypi-friendly-readme.rst]: And of course, re-reading the changeset I see that you already link to it, which makes me more in favor of the minimalist approach.

Comment (ddbeck) [reply source/guides/making-a-pypi-friendly-readme.rst]: @jonparrott please don't change my words to say something I would not have argued and then put it in a block quote. This feels a half step removed from the mocking SpongeBob meme.

Anyway, I'm chaffing a bit at the suggestion that my recommendations were arbitrary, so I'm pushing a change to drop the README content section entirely. Accept the PR or don't, but I'm done being needled over this.

**Comment by jonparrott** [comment]
FYI: chatted with @ddbeck over email and I'm cool with these changes. Merging. :)

Thanks, @ddbeck!
